limiter_atomic_int64.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2016,2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ratelimit // import "go.uber.org/ratelimit"
+
+import (
+	"runtime"
+	"time"
+
+	"sync/atomic"
+)
+
+type atomicInt64Limiter struct {
+	//lint:ignore U1000 Padding is unused but it is crucial to maintain performance
+	// of this rate limiter in case of collocation with other frequently accessed memory.
+	prepadding [64]byte // cache line size = 64; created to avoid false sharing.
+	state      int64    // unix nanoseconds of the next permissions issue.
+	//lint:ignore U1000 Padding is unused but it is crucial to maintain performance
+	// of this rate limiter in case of collocation with other frequently accessed memory.
+	postpadding [56]byte // cache line size - state pointer size = 64 - 8; created to avoid false sharing.
```

Comment:
Could you explain this one a bit more? Since we have other fields of the struct below, how does postpadding work here exactly?

Reply:
Every time when we perform CAS on `state` field, we mark the whole cache line as "dirty". So when some other goroutine wants to read `perRequest`, `maxSlack` or `clock` caching protocol will force us to reload the whole cache line even though  `perRequest`, `maxSlack` or `clock` are constants in our case. So we introduce `postpadding` to move `perRequest`, `maxSlack` and `clock` or separate cache line. More about it here https://en.wikipedia.org/wiki/False_sharing

Benchmarks that show perf degradations when we remove `postpadding` check that atomic_int64_no_padding and atomic_int64_no_sched_no_padding are slower than their counterparts.

<img width="1048" alt="Screenshot 2022-06-05 at 16 13 08" src="https://user-images.githubusercontent.com/3532750/172057409-22db3826-a353-4366-b8e1-22a0bc3638c7.png">

Reply:
Fascinating. This sounds like an even more advanced case of https://pkg.go.dev/golang.org/x/tools/go/analysis/passes/fieldalignment, where we want to separate read/write variables.

Questions:
(1) do you know if there's any runtime variable that defines cache size line and int64 line? Perhaps we could avoid the hardcodes ints, and thus make the code self documenting.
(2) The comment needs updating, it's no longer "statePointerSize"

Unrelated:
(3) Are you using same padding techniques in some other projects? Could you paste on some links? I'd be curious to see how often, and to what effect these are used.

Reply:
> Do you know if there's any runtime variable that defines cache size line and int64 line?

I didn't know about it before you asked, but apparently `golang.org/x/sys/cpu` package has `CacheLinePad` struct for it https://pkg.go.dev/golang.org/x/sys/cpu#CacheLinePad
I can you that, but I'd prefer to add it in a separate PR if you don't mind. I wanted to keep this one as is, not to remeasure everything from scratch.

Reply:
> The comment needs updating, it's no longer "statePointerSize"

Done

Reply:
> Are you using same padding techniques in some other projects? Could you paste on some links?

Yes, it is a widely used technic. I can find some good examples and post them here if you want. In some languages, like Java, for example, AtomicReference or AtomicInteger already have padding built-in.